File: src/lib/pictureInPicture.ts
/**
 * Exodus Browser — Picture-in-Picture (画中画) functionality
 * Frontend integration for PiP Tauri commands
 * Supports video element integration similar to Chrome/Firefox
 */

import { invoke } from '@tauri-apps/api/core';

export type PipState = {
  video_url: string;
  is_active: boolean;
  window_width: number;
  window_height: number;
};

/**
 * Enter Picture-in-Picture mode for a video
 */
export async function pipEnter(
  video_url: string,
  width: number = 640,
  height: number = 480
): Promise<void> {
  return invoke('pip_enter', { videoUrl: video_url, width, height });
}

/**
 * Exit Picture-in-Picture mode for a video
 */
export async function pipExit(video_url: string): Promise<void> {
  return invoke('pip_exit', { videoUrl: video_url });
}

/**
 * Resize Picture-in-Picture window
 */
export async function pipResize(
  video_url: string,
  width: number,
  height: number
): Promise<void> {
  return invoke('pip_resize', { videoUrl: video_url, width, height });
}

/**
 * Get Picture-in-Picture state for a specific video
 */
export async function pipGetState(video_url: string): Promise<PipState | null> {
  return invoke('pip_get_state', { videoUrl: video_url });
}

/**
 * Get all active Picture-in-Picture states
 */
export async function pipGetAllActive(): Promise<PipState[]> {
  return invoke('pip_get_all_active');
}

/**
 * Inject Picture-in-Picture API into a webview
 * This adds the HTML5 Picture-in-Picture API to the webview context
 * Security: Validates video URLs and sanitizes inputs before injection
 */
export function injectPictureInPictureAPI(webview: any): void {
  const script = `
    (function() {
      // Security: Check if already injected to prevent duplicate injections
      if (window.__EXODUS_PIP_INJECTED__) {
        console.log('Picture-in-Picture API already injected, skipping');
        return;
      }
      window.__EXODUS_PIP_INJECTED__ = true;
      
      // Security: Validate URL function
      function isValidVideoUrl(url) {
        if (!url || typeof url !== 'string') return false;
        try {
          const parsed = new URL(url);
          return parsed.protocol === 'http:' || 
                 parsed.protocol === 'https:' || 
                 parsed.protocol === 'data:';
        } catch {
          return false;
        }
      }
      
      // Security: Validate dimensions
      function isValidDimensions(width, height) {
        const minSize = 160;
        const maxSize = 4096;
        return width >= minSize && width <= maxSize && 
               height >= minSize && height <= maxSize;
      }
      
      // Override requestPictureInPicture on HTMLVideoElement prototype
      if (HTMLVideoElement.prototype.requestPictureInPicture) {
        const originalRequestPiP = HTMLVideoElement.prototype.requestPictureInPicture;
        
        HTMLVideoElement.prototype.requestPictureInPicture = async function() {
          const video = this;
          const videoUrl = video.src || video.currentSrc;
          const width = video.videoWidth || 640;
          const height = video.videoHeight || 480;
          
          // Security: Validate inputs
          if (!isValidVideoUrl(videoUrl)) {
            console.error('Invalid video URL for Picture-in-Picture:', videoUrl);
            throw new Error('Invalid video URL');
          }
          
          if (!isValidDimensions(width, height)) {
            console.error('Invalid dimensions for Picture-in-Picture:', width, height);
            throw new Error('Invalid dimensions');
          }
          
          // Call Tauri backend
          if (window.__TAURI__) {
            try {
              await window.__TAURI__.invoke('pip_enter', {
                videoUrl: videoUrl,
                width: width,
                height: height
              });
              
              // Return a mock PiP window object
              return {
                onresize: null,
                onenter: null,
                onleave: null,
                width: width,
                height: height
              };
            } catch (error) {
              console.error('Picture-in-Picture failed:', error);
              throw error;
            }
          }
          
          // Fallback to native API if available
          return originalRequestPiP.call(video);
        };
      }
      
      // Add exitPictureInPicture if not present
      if (!HTMLVideoElement.prototype.exitPictureInPicture) {
        HTMLVideoElement.prototype.exitPictureInPicture = async function() {
          const video = this;
          const videoUrl = video.src || video.currentSrc;
          
          // Security: Validate URL
          if (!isValidVideoUrl(videoUrl)) {
            console.error('Invalid video URL for Picture-in-Picture exit:', videoUrl);
            throw new Error('Invalid video URL');
          }
          
          if (window.__TAURI__) {
            try {
              await window.__TAURI__.invoke('pip_exit', {
                videoUrl: videoUrl
              });
              return;
            } catch (error) {
              console.error('Exit Picture-in-Picture failed:', error);
              throw error;
            }
          }
        };
      }
      
      // Listen for Tauri events to update video state
      if (window.__TAURI__) {
        window.__TAURI__.event.listen('exodus-pip-entered', (event) => {
          console.log('Picture-in-Picture entered:', event.payload);
        });
        
        window.__TAURI__.event.listen('exodus-pip-exited', (event) => {
          console.log('Picture-in-Picture exited:', event.payload);
        });
        
        window.__TAURI__.event.listen('exodus-pip-resized', (event) => {
          console.log('Picture-in-Picture resized:', event.payload);
        });
      }
      
      console.log('Picture-in-Picture API injected successfully');
    })();
  `;
  
  webview.executeJavaScript(script);
}

File: src-tauri/src/picture_in_picture.rs
//! Picture-in-Picture (画中画) functionality for Exodus Browser
//! Supports HTML5 video elements with Picture-in-Picture API

use std::sync::Arc;
use tauri::{AppHandle, Emitter};
use tokio::sync::Mutex;

const MIN_WINDOW_SIZE: u32 = 160;
const MAX_WINDOW_SIZE: u32 = 4096;
const MAX_URL_LENGTH: usize = 2048;

/// Validate URL format
pub fn validate_url(url: &str) -> Result<(), String> {
    if url.is_empty() {
        return Err("URL cannot be empty".to_string());
    }
    
    if url.len() > MAX_URL_LENGTH {
        return Err(format!("URL too long (max {} characters)", MAX_URL_LENGTH));
    }
    
    // Basic URL validation
    if !url.starts_with("http://") && !url.starts_with("https://") && !url.starts_with("data:") {
        return Err("URL must start with http://, https://, or data:".to_string());
    }
    
    Ok(())
}

/// Validate window dimensions
pub fn validate_dimensions(width: u32, height: u32) -> Result<(), String> {
    if width < MIN_WINDOW_SIZE || width > MAX_WINDOW_SIZE {
        return Err(format!("Width must be between {} and {}", MIN_WINDOW_SIZE, MAX_WINDOW_SIZE));
    }
    
    if height < MIN_WINDOW_SIZE || height > MAX_WINDOW_SIZE {
        return Err(format!("Height must be between {} and {}", MIN_WINDOW_SIZE, MAX_WINDOW_SIZE));
    }
    
    Ok(())
}

/// Picture-in-Picture state for a specific video
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct PipState {
    pub video_url: String,
    pub is_active: bool,
    pub window_width: u32,
    pub window_height: u32,
}

/// Picture-in-Picture manager
#[derive(Clone)]
pub struct PipManager {
    states: Arc<Mutex<Vec<PipState>>>,
}

impl PipManager {
    pub fn new() -> Self {
        Self {
            states: Arc::new(Mutex::new(Vec::new())),
        }
    }

    /// Enter Picture-in-Picture mode for a video
    pub async fn enter_pip(&self, video_url: String, width: u32, height: u32, app: AppHandle) -> Result<(), String> {
        // Validate inputs
        validate_url(&video_url)?;
        validate_dimensions(width, height)?;
        
        let mut states = self.states.lock().await;
        
        // Check if this video is already in PiP
        if let Some(existing) = states.iter().find(|s| s.video_url == video_url) {
            if existing.is_active {
                return Ok(()); // Already active
            }
        }
        
        // Create new PiP state
        let state = PipState {
            video_url: video_url.clone(),
            is_active: true,
            window_width: width,
            window_height: height,
        };
        
        // Remove old entry for this video if exists
        states.retain(|s| s.video_url != video_url);
        states.push(state.clone());
        
        // Emit event
        let _ = app.emit("exodus-pip-entered", &state);
        
        Ok(())
    }

    /// Exit Picture-in-Picture mode for a video
    pub async fn exit_pip(&self, video_url: String, app: AppHandle) -> Result<(), String> {
        let mut states = self.states.lock().await;
        
        if let Some(index) = states.iter().position(|s| s.video_url == video_url) {
            let mut state = states[index].clone();
            state.is_active = false;
            states[index] = state.clone();
            
            let _ = app.emit("exodus-pip-exited", &state);
        }
        
        Ok(())
    }

    /// Resize Picture-in-Picture window
    pub async fn resize_pip(&self, video_url: String, width: u32, height: u32, app: AppHandle) -> Result<(), String> {
        // Validate inputs
        validate_url(&video_url)?;
        validate_dimensions(width, height)?;
        
        let mut states = self.states.lock().await;
        
        if let Some(state) = states.iter_mut().find(|s| s.video_url == video_url) {
            state.window_width = width;
            state.window_height = height;
            
            let _ = app.emit("exodus-pip-resized", state.clone());
        } else {
            return Err("Picture-in-Picture window not found for this video".to_string());
        }
        
        Ok(())
    }

    /// Get Picture-in-Picture state for a specific video
    pub async fn get_pip_state(&self, video_url: String) -> Option<PipState> {
        let states = self.states.lock().await;
        states.iter().find(|s| s.video_url == video_url).cloned()
    }
    
    /// Get all active Picture-in-Picture states
    pub async fn get_all_active(&self) -> Vec<PipState> {
        let states = self.states.lock().await;
        states.iter().filter(|s| s.is_active).cloned().collect()
    }
}

impl Default for PipManager {
    fn default() -> Self {
        Self::new()
    }
}

// Tauri commands for Picture-in-Picture

/// Enter Picture-in-Picture mode
#[tauri::command]
pub async fn pip_enter(
    video_url: String,
    width: u32,
    height: u32,
    pip_manager: tauri::State<'_, Arc<PipManager>>,
    app: AppHandle,
) -> Result<(), String> {
    pip_manager.enter_pip(video_url, width, height, app).await
}

/// Exit Picture-in-Picture mode
#[tauri::command]
pub async fn pip_exit(
    video_url: String,
    pip_manager: tauri::State<'_, Arc<PipManager>>,
    app: AppHandle,
) -> Result<(), String> {
    pip_manager.exit_pip(video_url, app).await
}

/// Resize Picture-in-Picture window
#[tauri::command]
pub async fn pip_resize(
    video_url: String,
    width: u32,
    height: u32,
    pip_manager: tauri::State<'_, Arc<PipManager>>,
    app: AppHandle,
) -> Result<(), String> {
    pip_manager.resize_pip(video_url, width, height, app).await
}

/// Get Picture-in-Picture state for a video
#[tauri::command]
pub async fn pip_get_state(
    video_url: String,
    pip_manager: tauri::State<'_, Arc<PipManager>>,
) -> Result<Option<PipState>, String> {
    Ok(pip_manager.get_pip_state(video_url).await)
}

/// Get all active Picture-in-Picture states
#[tauri::command]
pub async fn pip_get_all_active(
    pip_manager: tauri::State<'_, Arc<PipManager>>,
) -> Result<Vec<PipState>, String> {
    Ok(pip_manager.get_all_active().await)
}
